classes/access_manager.php

```diff
@@ -0,0 +1,147 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Manage the access to the quiz.
+ *
+ * @package    quizaccess_seb
+ * @author     Tim Hunt
+ * @author     Luca Bosch
+ * @author     Andrew Madden <andrewmadden@catalyst-au.net>
+ * @copyright  2019 Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace quizaccess_seb;
+
+use context_module;
+use quiz;
+
+defined('MOODLE_INTERNAL') || die();
+
+class access_manager {
+
+    private const CONFIG_KEY_HEADER = 'HTTP_X_SAFEEXAMBROWSER_CONFIGKEYHASH';
+
+    private const BROWSER_EXAM_KEY_HEADER = 'HTTP_X_SAFEEXAMBROWSER_REQUESTHASH';
+
+    /** @var quiz $quiz A quiz object containing all information pertaining to current quiz. */
+    private $quiz;
+
+    /** @var quiz_settings $quizsettings A quiz settings persistent object containing plugin settings */
+    private $quizsettings;
+
+    /** @var context_module $context Context of this quiz activity. */
+    private $context;
+
+    /**
+     * The access_manager constructor.
+     *
+     * @param quiz $quiz The details of the quiz.
+     *
+     * @throws \moodle_exception
+     */
+    public function __construct(quiz $quiz) {
+        $this->quiz = $quiz;
+        $this->context = context_module::instance($quiz->get_cmid());
+        $this->quizsettings = quiz_settings::get_record(['quizid' => $quiz->get_quizid()]);
+        if ($this->quizsettings == false) {
+            throw new \moodle_exception('noconfigfound', 'quizaccess_seb', '', $quiz->get_cmid());
+        }
+    }
+
+    /**
+     * Check if the current request contains the required key hashes in header.
+     *
+     * @return bool true if the user is using a browser with a permitted key, false if not.
+     *
+     * @throws \coding_exception
+     */
+    public function validate_access_keys() : bool {
+        $configkey = $this->quizsettings->get('configkey');
+        if (empty($configkey)) {
+            return false;
+        }
+
+        // If the Config Key header isn't present, prevent access.
+        if (!array_key_exists(self::CONFIG_KEY_HEADER, $_SERVER)) {
+            return false;
+        }
+
+        return self::check_key($configkey, self::get_this_page_url(),
```

Comment:
why calling with self:: ?and not $this->

Reply:
Copy-paste error. Fixed